homework/zad9_opencv/line_match2.cpp

```diff
@@ -0,0 +1,161 @@
+#include <opencv2/core.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/highgui.hpp>
+
+#include <iostream>
+#include <mutex>
+#include <future>
+#include <thread>
+
+void show_wait_destroy(const char* winname, cv::Mat img) {
+    cv::imshow(winname, img);
+    cv::moveWindow(winname, 500, 0);
+    cv::waitKey(0);
+    cv::destroyWindow(winname);
+}
+
+cv::Mat createMatLineAlpha( const cv::Size& size, int iAlpha )
+{
+    cv::Mat mat(size, CV_8U);
+    cv::Point origin( size.width / 2, size.height / 2 );
+    double dRad = ( iAlpha * M_PI ) / 180.0;
+
+    if ( iAlpha != 90 )
+    {
+        for( int iCol = 0; iCol < size.width; ++iCol )
+        {
+            double dValue = std::tan(dRad) * ( iCol - origin.x );
+            int iPos = origin.y - std::lround(dValue);
+
+            if ( iPos >= 0 && iPos < size.height )
+            {
+                mat.at<unsigned char>( iPos , iCol) = 1;
+            }
+        }
+    }
+    else
+    {
+        for ( int iRow = 0; iRow < size.height; ++iRow )
+        {
+            mat.at<unsigned char>(iRow, origin.x) = 1;
+        }
+    }
+
+    return mat;
+}
+
+auto lineMatchFunction = [](const cv::Mat& bw_image, const cv::Size& size, int alpha)
+{
+    // Create structure element for extracting lines through morphology operations
+    cv::Mat imgStructure = createMatLineAlpha(size, alpha);
+    cv::Mat imgLocal = bw_image.clone();
+
+    // Apply morphology operations
+    cv::erode(imgLocal, imgLocal, imgStructure, cv::Point(-1, -1));
+    cv::dilate(imgLocal, imgLocal, imgStructure, cv::Point(-1, -1));
+
+    return imgLocal.clone();
```

Comment:
nie będzie lepiej bez clone() przy return? Dodatkowa kopia chyba nam nic nie daje i może powodować, że RVO nie zostanie zastosowane.

Reply:
To byl eksperyment zwiazany z moimi podejrzeniami co do funkcji "clone". 
Zobacz moje wczesniejsze uwagi.